packages/webview_flutter/webview_flutter_wkwebview/lib/src/web_kit/web_kit.dart

```diff
@@ -27,10 +42,149 @@ enum WKAudiovisualMediaType {
   all,
 }
 
+/// A script that the web view injects into a webpage.
+///
+/// Wraps [WKUserScript](https://developer.apple.com/documentation/webkit/wkuserscript?language=objc).
+class WKUserScript {
+  /// Constructs a [UserScript].
+  WKUserScript(
+    this.source,
+    this.injectionTime, {
+    required this.isMainFrameOnly,
+  });
+
+  /// The script’s source code.
+  final String source;
+
+  /// The time at which to inject the script into the webpage.
+  final WKUserScriptInjectionTime injectionTime;
+
+  /// Indicates whether to inject the script into the main frame or all frames.
+  final bool isMainFrameOnly;
+}
+
+/// An object that encapsulates a message sent by JavaScript code from a webpage.
+///
+/// Wraps [WKScriptMessage](https://developer.apple.com/documentation/webkit/wkscriptmessage?language=objc).
+class WKScriptMessage {
+  /// Constructs a [WKScriptMessage].
+  WKScriptMessage({required this.name, this.body});
+
+  /// The name of the message handler to which the message is sent.
+  final String name;
+
+  /// The body of the message.
+  ///
+  /// Allowed types are [num], [String], [List], [Map], and `null`.
+  final Object? body;
+}
+
+/// An interface for receiving messages from JavaScript code running in a webpage.
+///
+/// Wraps [WKScriptMessageHandler](https://developer.apple.com/documentation/webkit/wkscriptmessagehandler?language=objc)
+class WKScriptMessageHandler {
```

Comment:
I like this idea since it would make this simpler and easier to test. But my main concern would be that this protocol could eventually be updated with a new method, then we would have to convert it back to a class. Although this would be relatively trivial to do; ~50 line change.

So I wonder if you think it's better to play it safe and avoid larger changes in the future or use our intuition to convert one-method classes/protocols to typedefs. Java uses a lot of one method interfaces, so this policy could definitely apply to the Android code as well.

Reply:
I was thinking that since it was entirely internal it wouldn't be a big deal if we had to convert it to a class (if it were public I'd definitely say we should keep this version). But I guess in cases where we did have to do that conversion later it would undermine your design goal of this approach of putting the language boundary as close to the OS APIs as possible in order to make updating it for small OS API changes a simple, mechanical change.

I guess it really becomes a gamble about whether we think new methods are likely or not. It's tricky since the pattern of using protocols for delegation, rather than blocks, is baked pretty deeply into ObjC norms, so the existence of a protocol isn't really an indication of whether the designers expect to add more methods.

I'll leave it to you to decide since I think there are good arguments both ways here, and the main deciding factor is unknowable.

Reply:
My preference is to play it safe so that it can remain easy to maintain for future maintainers and community developers. So I'll keep it as is for now.